Marketplace: Remove variations.product_slug dependency from plugins

epeicher opened the pull request:
> **Note**: Depends on D87698-code

#### Proposed Changes

The `plugin` object contains a `variations` object that has the form 
```
// plugin.variations
{
    monthly: { product_slug: <monthly_product_slug>, product_id: <monthly_product_id> },
    yearly: { product_slug: <yearly_product_slug>, product_id: <yearly_product_id> },
}
```
This field is populated from the backend for `wpcom` plugins so it needs to be stored and indexed by the backend. Those fields are used to obtain the prices for each period.

To avoid returning both fields, `product_id` and `product_slug`, this change will calculate `product_slug` from the list of products using the `product_id` so `product_slug` can be removed from the backend.

The backend changes have been applied as part of D87698-code

#### Testing Instructions

The changes should be tested with and _without_ the related diff D87698-code applied, as this front-end change must be applied **before** the backend change

<!--
Add as many details as possible to help others reproduce the issue and test the fix.
"Before / After" screenshots can also be very helpful when the change is visual.
-->

1. Using the production API, i.e. **not** being sandboxed, apply this branch
2. Navigate to a _paid_ plugin details page, e.g. `/plugins/woocommerce-subscriptions/<site_id>` being `<site_id>` any of your sites url
3. Make sure you can see the correct prices for `Monthly` and `Annually` periods
4. Make sure you can see the discount 
![CleanShot 2022-09-15 at 11 10 45@2x](https://user-images.githubusercontent.com/3519124/190364659-7c1b685d-a61f-4a61-b5e2-dfbae78a1e2c.png)
8. Make sure you can add the product to the cart and you can see the correct product added
5. Navigate to a free plugin, e.g. `/plugins/zero-bs-crm/<site_id>` or select one from the list of plugins
6. Make sure you do not see any unexpected errors
8. Sandbox the API and apply the following diff D87698-code
9. Repeat the steps 2 to 7 to make sure there are no unexpected errors 
10. Additionally, use `jetpack-search` in some testing to check there are no unexpected errors as that plugin is a special case of preinstalled premium plugin 

#### Pre-merge Checklist

<!--
Complete applicable items on this checklist **before** merging into trunk. Inapplicable items can be left unchecked.

Both the PR author and reviewer are responsible for ensuring the checklist is completed.
-->

- [ ] [Have you written new tests](https://wpcalypso.wordpress.com/devdocs/docs/testing/index.md) for your changes?
- [ ] Have you tested the feature in Simple (P9HQHe-k8-p2), Atomic (P9HQHe-jW-p2), and self-hosted Jetpack sites (PCYsg-g6b-p2)?
- [ ] Have you checked for TypeScript, React or other console errors?
- [ ] Have you used memoizing on expensive computations? More info in [Memoizing with create-selector](https://github.com/Automattic/wp-calypso/blob/trunk/packages/state-utils/src/create-selector/README.md) and [Using memoizing selectors](https://react-redux.js.org/api/hooks#using-memoizing-selectors) and [Our Approach to Data](https://github.com/Automattic/wp-calypso/blob/trunk/docs/our-approach-to-data.md)
- [ ] Have we added the "[Status] String Freeze" label as soon as any new strings were ready for translation (p4TIVU-5Jq-p2)?

<!--
Link a related issue to this PR. If the PR does not immediately resolve the issue,
for example, it requires a separate deployment to production, avoid
using the "fixes" keyword and instead attach the [Status] Fix Inbound label to
the linked issue.
-->

Fixes #67507

epeicher committed:
Marketplace: Remove variations.product_slug dependency

This field can be obtained from the product list and removing it will
avoid storing and returning that field by the backend

epeicher commented:
While doing an E2E testing and finding `variations` in the whole codebase I have found that is used in some other components and removing `variation.product_slug` will cause some regressions with the current change, for example the discount is not displayed with the current change
![CleanShot 2022-09-14 at 12 17 03@2x](https://user-images.githubusercontent.com/3519124/190128561-14ac9343-c642-4cad-8d45-5318fd7b971e.png)

I am going to try a different approach: Update the `variations.plugin_slug` just after being populated from the backend so every _consumer_ can be unchanged.

gcsecsey commented:
> I am going to try a different approach: Update the variations.plugin_slug just after being populated from the backend so every consumer can be unchanged.

This sounds like a good approach! 👍

epeicher committed:
Update all usages of variation.product_slug to use product_id

All the usages of variation.product_slug must be updated to use a new
function added to utils. That function will use product_id if
product_slug does not exist

epeicher commented:
As part of https://github.com/Automattic/wp-calypso/pull/67787/commits/21fa8cf3ede5ccdc25c0bf2cdffa91753d2d3e06 a new function `getProductSlugByPeriodVariation` have been added to a `lib/plugins/utils` package. That function will be used every time the `variations.plugin_slug` is required. 

I have tried populating the `plugin.variations` with an object that contains `product_slug` as mentioned in [my previous comment](https://github.com/Automattic/wp-calypso/pull/67787#issuecomment-1246555567) but I needed to pass the `productsList` to the `useWPCOMPlugin` ([link here](https://github.com/Automattic/wp-calypso/blob/21fa8cf3ede5ccdc25c0bf2cdffa91753d2d3e06/client/data/marketplace/use-wpcom-plugins-query.ts#L81-L82)) hook which seemed like unnecessary coupling but I am open to suggestions.

epeicher commented:
The test steps have been updated to reflect the new changes. The task is ready for a review

cc @cpapazoglou

gcsecsey commented:
I've tested the Calypso changes both with and without the backend patch, and everything works as described.

Changes LGTM! 🚢
